   # Evolution Case Study

## 1. Introduction

### Purpose of the Analysis

The objective of this examination is to extract meaningful observations from the data supplied by Evolution, specifically focusing on the Casino Data and Blackjack Data files. The goal is to pinpoint the crucial cities for the company and present a condensed overview of noteworthy findings, accompanied by suggested metrics for assessing business performance.

### Data Sources

The data comes from Lipstick Casino.

1. Casino Data
2. Blackjack Data

## 2. Setup and Tools

The examination was carried out employing Power Query, PostgreSQL, Tableau, and finally Jupyter Notebook for documentation.

Power Query was utilized for fundamental modifications, including the alteration of column names and data types.
PostgreSQL was employed to delve deeper into the data, deriving insights to inform decisions before visualization construction.
Tableau was used to create visualizations aimed at gaining insights into player behavior and enhancing our business decision-making processes.

In [2]:
import psycopg2
from sqlalchemy import create_engine, text

In [4]:
import pandas as pd

In [17]:
from IPython.display import Image

In [3]:
db_params = {
    'host': 'localhost',
    'database': 'Evolution',
    'user': 'postgres',
    'password': '2168165567'
}

In [5]:
db_params['database'] = 'Evolution'
engine = create_engine(f'postgresql://{db_params["user"]}:{db_params["password"]}@{db_params["host"]}/{db_params["database"]}')

## Define the file path for your CSV files
csv_files = {
    'casino': r"C:\Users\mmoki\OneDrive\Desktop\Evolution\casino.csv",
    'blackjack': r"C:\Users\mmoki\OneDrive\Desktop\Evolution\blackjack.csv"
}

In [7]:
# Loop through the CSV files and import them into PostgreSQL
for table_name, file_path in csv_files.items():
    df = pd.read_csv(file_path)
    df.to_sql(table_name, engine, if_exists='replace', index=False)

Now that we are done with connecting jupyter notebook and postgeSQL, lets do some more EDA.

In [8]:
%load_ext sql

%sql postgresql://postgres:2168165567@localhost:5432/Evolution

In [9]:
%%sql

SELECT * 
FROM casino
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/Evolution
10 rows affected.


[('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'Baccarat A', 'EUR', 'January', 2, 85, 4, 4),
 ('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'Baccarat B', 'EUR', 'January', -78, 184, 3, 3),
 ('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'Baccarat B', 'EUR', 'March', -35, 35, 1, 1),
 ('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'Baccarat C', 'EUR', 'January', -25, 25, 1, 1),
 ('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'First Person Baccarat', 'EUR', 'January', 2, 2, 2, 2),
 ('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'First Person Baccarat', 'EUR', 'July', 0, 2, 1, 2),
 ('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'Speed Baccarat A', 'EUR', 'January', 169, 1163, 35, 35),
 ('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'Speed Baccarat A', 'EUR', 'March', -269, 9351, 1137, 1169),
 ('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'Speed Baccarat A', 'EUR', 'June', -428, 4705, 191, 191),
 ('7/1/2020', 'Lipstick Casino', 'Estonia', 'Baccarat', 'Other', 'Speed Baccarat A', 'EUR', 'July', 0, 6, 3, 3)]

In [10]:
%%sql

SELECT * 
FROM blackjack
LIMIT 10;

 * postgresql://postgres:***@localhost:5432/Evolution
10 rows affected.


[('lipstick_casino', '165f9a23e4702d74d8904222', 'blackjack', '2/1/2021 10:51', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_213Seat3', 'Helsinki', 'SmartPhone', 'iOS', 6.0, 0, 'robinyy', 0.0984),
 ('lipstick_casino', '165f9a23e4702d74d8904222', 'blackjack', '2/1/2021 10:51', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_PerfectPairSeat3', 'Helsinki', 'SmartPhone', 'iOS', 4.0, 0, 'robinyy', 0.0984),
 ('lipstick_casino', '165f99e52ae31f2c31dbfa5a', 'blackjack', '2/1/2021 10:46', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_213Seat3', 'Helsinki', 'SmartPhone', 'iOS', 5.0, 0, 'robinyy', 0.0984),
 ('lipstick_casino', '165f99e52ae31f2c31dbfa5a', 'blackjack', '2/1/2021 10:46', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_PerfectPairSeat3', 'Helsinki', 'SmartPhone', 'iOS', 5.0, 0, 'robinyy', 0.0984),
 ('lipstick_casino', '165f9a13d86d2b9c4f6d55e7', 'blackjack', '2/1/2021 10:50', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_213Seat3', 'Helsinki', 'SmartPhone', 'iOS', 3.0, 0, 'robinyy', 0.0984),
 ('lipstick_casino', '165f9a13d86d2b9c4f6d55e7', 'blackjack', '2/1/2021 10:50', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_PerfectPairSeat3', 'Helsinki', 'SmartPhone', 'iOS', 2.0, 0, 'robinyy', 0.0984),
 ('lipstick_casino', '165f9a06b02afeed9bd69aa2', 'blackjack', '2/1/2021 10:49', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_213Seat3', 'Helsinki', 'SmartPhone', 'iOS', 5.0, 0, 'robinyy', 0.0984),
 ('lipstick_casino', '165f9a06b02afeed9bd69aa2', 'blackjack', '2/1/2021 10:49', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_PerfectPairSeat3', 'Helsinki', 'SmartPhone', 'iOS', 5.0, 0, 'robinyy', 0.0984),
 ('lipstick_casino', '165f9a30ec7358d0c4b07a8b', 'blackjack', '2/1/2021 10:52', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_213Seat3', 'Helsinki', 'SmartPhone', 'iOS', 5.0, 30, 'robinyy', 0.0984),
 ('lipstick_casino', '165f9a30ec7358d0c4b07a8b', 'blackjack', '2/1/2021 10:52', 'luw6zjxjuarambhh', 'EUR', 'EUR', '3tg4f8vpj58soi9d', 'BJ_PerfectPairSeat3', 'Helsinki', 'SmartPhone', 'iOS', 5.0, 0, 'robinyy', 0.0984)]

In [13]:
%%sql

SELECT 
country,
COUNT(player_game_count) AS amount_of_players
FROM casino
GROUP BY 1

 * postgresql://postgres:***@localhost:5432/Evolution
4 rows affected.


[('Finland', 1152), ('Sweden', 1130), ('Norway', 17), ('Estonia', 843)]

## 3. Data Exploration

Examining the 'Casino Data' reveals information pertaining to four countries: Sweden, Estonia, Finland, and Norway. Notable statistics indicate that Finland boasts the highest player count with 1152, closely followed by Sweden with 1130, Estonia with 843, and Norway with 17 players. Shifting attention to the temporal aspect, the data under scrutiny was gathered between July 1, 2020, and December 1, 2020.

On the other hand, the 'Blackjack Data' predominantly centers around 89 players hailing from 37 cities. This dataset covers the span of a single day, providing insights into the activity of players throughout the day.


## Table 1


In [26]:
%%sql

WITH players AS (
  SELECT
    DATE_TRUNC('month', CAST(month_year AS TIMESTAMP)) AS month,
    country,
    COUNT(player_game_count) AS monthly_players
  FROM
    casino
  GROUP BY
    1,
    2
)
SELECT
  month,
  country,
  monthly_players,
  monthly_players - LAG(monthly_players) OVER (PARTITION BY country ORDER BY month) AS player_difference
FROM
  players;


 * postgresql://postgres:***@localhost:5432/Evolution
24 rows affected.


[(datetime.datetime(2020, 7, 1, 0, 0), 'Estonia', 115, None),
 (datetime.datetime(2020, 8, 1, 0, 0), 'Estonia', 121, 6),
 (datetime.datetime(2020, 9, 1, 0, 0), 'Estonia', 140, 19),
 (datetime.datetime(2020, 10, 1, 0, 0), 'Estonia', 134, -6),
 (datetime.datetime(2020, 11, 1, 0, 0), 'Estonia', 150, 16),
 (datetime.datetime(2020, 12, 1, 0, 0), 'Estonia', 183, 33),
 (datetime.datetime(2020, 7, 1, 0, 0), 'Finland', 157, None),
 (datetime.datetime(2020, 8, 1, 0, 0), 'Finland', 167, 10),
 (datetime.datetime(2020, 9, 1, 0, 0), 'Finland', 182, 15),
 (datetime.datetime(2020, 10, 1, 0, 0), 'Finland', 193, 11),
 (datetime.datetime(2020, 11, 1, 0, 0), 'Finland', 212, 19),
 (datetime.datetime(2020, 12, 1, 0, 0), 'Finland', 241, 29),
 (datetime.datetime(2020, 7, 1, 0, 0), 'Norway', 8, None),
 (datetime.datetime(2020, 8, 1, 0, 0), 'Norway', 2, -6),
 (datetime.datetime(2020, 9, 1, 0, 0), 'Norway', 1, -1),
 (datetime.datetime(2020, 10, 1, 0, 0), 'Norway', 1, 0),
 (datetime.datetime(2020, 11, 1, 0, 0), 'Norway', 1, 0),
 (datetime.datetime(2020, 12, 1, 0, 0), 'Norway', 4, 3),
 (datetime.datetime(2020, 7, 1, 0, 0), 'Sweden', 144, None),
 (datetime.datetime(2020, 8, 1, 0, 0), 'Sweden', 154, 10),
 (datetime.datetime(2020, 9, 1, 0, 0), 'Sweden', 171, 17),
 (datetime.datetime(2020, 10, 1, 0, 0), 'Sweden', 197, 26),
 (datetime.datetime(2020, 11, 1, 0, 0), 'Sweden', 230, 33),
 (datetime.datetime(2020, 12, 1, 0, 0), 'Sweden', 234, 4)]

![Dashboard1](Dashboard1.png)

For this dashboard, my primary emphasis is on specific metrics. As previously stated, two currencies EUR and SEK, were employed, and the stacked bar chart illustrates monthly wager and GGR in both currencies. Regarding the month-to-month player visualization, due to the absence of player_id information, I assumed: a Player Game Count of 1 signifies a new player for each corresponding row, while counts greater than 1 imply player retention.  Prior to building I used SQL to make sure I got the results needed. 

Link to the dashboard - https://public.tableau.com/app/profile/mokonen.michael6645/viz/Casino1_17049161591230/Dashboard1#3

## Table 2.

![Dashboard2](Dashboard2.png)


## Table 3

In [36]:
%%sql

WITH time_period AS (
  SELECT
    CAST(game_start AS TIME) AS game_time,
    customer_screen_name,
    city,
    wager_eur,
    payout_eur
  FROM blackjack
),
timing AS (
  SELECT
    customer_screen_name,
    city,
    CASE
      WHEN game_time >= '00:00:00' AND game_time < '12:00:00' THEN 'Morning'
      ELSE 'Evening'
    END AS time_of_day
  FROM time_period
)
SELECT
  COUNT(customer_screen_name) AS players_prefer,
  time_of_day
FROM timing
GROUP BY 2;

 * postgresql://postgres:***@localhost:5432/Evolution
2 rows affected.


[(1486, 'Evening'), (673, 'Morning')]

![Dashboard3](Table3.png)

The analysis of player behavior reveals a trend where player activity and casino payouts are notably higher during the evening hours. This pattern is further supported by the examination of total payout amounts. To enhance our understanding of player preferences, we considered the devices used to access our games. The majority of players prefer smartphones, followed by desktops and tablets. This suggests an opportunity to enhance game features for smartphone users and consider mobile-friendly designs for future games, aligning with the prevalent player behavior.

link = https://public.tableau.com/app/profile/mokonen.michael6645/viz/MorningvsEvening/Dashboard1

## Table 4

In [37]:
%%sql

WITH
  games_played AS (
  SELECT
    country,
    game_category,
    COUNT(game_category) AS game_preference
  FROM
    casino
  GROUP BY
    1,
    2)
  SELECT
    country,
    game_category,
    RANK() OVER (PARTITION BY country ORDER BY game_preference DESC) AS top_ranked_games
  FROM
    games_played

 * postgresql://postgres:***@localhost:5432/Evolution
30 rows affected.


[('Estonia', 'Roulette', 1),
 ('Estonia', 'Blackjack', 2),
 ('Estonia', 'Poker', 3),
 ('Estonia', 'Baccarat', 4),
 ('Estonia', 'Game Shows', 5),
 ('Estonia', 'Money Wheel', 6),
 ('Estonia', 'Dice', 7),
 ('Estonia', 'Dragon Tiger', 8),
 ('Finland', 'Roulette', 1),
 ('Finland', 'Blackjack', 2),
 ('Finland', 'Poker', 3),
 ('Finland', 'Game Shows', 4),
 ('Finland', 'Baccarat', 5),
 ('Finland', 'Money Wheel', 6),
 ('Finland', 'Dice', 7),
 ('Finland', 'Dragon Tiger', 8),
 ('Norway', 'Blackjack', 1),
 ('Norway', 'Roulette', 2),
 ('Norway', 'Money Wheel', 2),
 ('Norway', 'Game Shows', 2),
 ('Norway', 'Poker', 5),
 ('Norway', 'Baccarat', 5),
 ('Sweden', 'Blackjack', 1),
 ('Sweden', 'Roulette', 2),
 ('Sweden', 'Baccarat', 3),
 ('Sweden', 'Poker', 4),
 ('Sweden', 'Dice', 5),
 ('Sweden', 'Money Wheel', 6),
 ('Sweden', 'Game Shows', 7),
 ('Sweden', 'Dragon Tiger', 8)]

![Dashboard4](Table4.png)

The presented heatmap provides insights into the popularity of games based on player participation in different countries. Notably, Norway stands out with a lower player count, while Estonia demonstrates robust engagement. Interestingly, the Money Wheel and Roulette games exhibit particularly strong performance, as evident from the heatmap.

Link = https://public.tableau.com/app/profile/mokonen.michael6645/viz/HeatMap_17049484912650/Sheet1

## Table 5 and 6

For this specific inquiry, my emphasis was primarily on leveraging the Blackjack dataset. Understanding player behavior, discerning city-wise trends, and evaluating the overall performance of the casino were pivotal aspects. While a more extensive data collection period would have provided a comprehensive view, analyzing the casino's performance during the specified day yielded valuable insights. Notably, the task necessitated the use of parameters and table calculations to construct the visualization, adding a layer of precision to the analysis.

![visualization5](ggr.png)

![visualization6](activevsinactive.png)

The parameter I created "Player Activity" which allows for a dynamic threshold setting. Instead of using a fized value in the calculation, it enables to interactively adjust the threshold that determines wheather a the players are considered "Active" or "Inactive" 

The table calculation is called 'NumPlays' field represents the count of plays for each player. 

Link - https://public.tableau.com/app/profile/mokonen.michael6645/viz/Table5and6/Sheet4

## Limitations

1. The Blackjack dataset only covers a single day, restricting the ability to observe trends over time. Acknowledge this limitation and interpret findings with caution, recognizing that a more extended data collection period would offer a more comprehensive understanding.
2. The inability to join the Casino and Blackjack datasets limits the depth of the analysis. Efforts should be made to address data integration challenges to provide a more unified view of player behavior across different games.